[CodeCompletion] Precompute "filter name"

rintaro opened the pull request:
"Filter name" (aka "name") is computed from a `CodeCompletionString`, that is not a trivial operation.
Previously, filter names are computed when they are used. Since they are always used, there's not much benefit to compute them lazily. Also some operations used to compute them, used it, then threw it away. Especially, for cached completions, filter names for all global symbols from imported modules are computed for each completion session.

In this PR, precompute and cache "filter name" in `ContextFreeCodeCompletionResult`. Getting "filter name" is now a trivial operation. And it's guaranteed that each "filter name" is computed only once.

rdar://84036006 rdar://89058389

rintaro committed:
[CodeCompletion] 'ContextFreeCodeCompletionResult' factory method

Convert 'ContextFreeCodeCompletionResult' constructor overloads to
'create()' factory methods. This is the consistent interface with
'CodeCompletionString'. NFC

rintaro committed:
[CodeCompletion] Precompute and cache "filter name"

Filter name for completion item is always used. Also, for cached items,
they are used multiple times for filtering. So precomputing and caching
it improves performance.

rdar://84036006

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please Sourcekit Stress test

rintaro committed:
[Basic] Move copyCString to Basic/StringExtras.

Also, use StringRef.copy() instead of copyString().

bnbarham commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -323,6 +309,63 @@ void CodeCompletionString::dump() const {
   }
 }
 
+ContextFreeCodeCompletionResult *ContextFreeCodeCompletionResult::create(
```
Would giving these more descriptive names make sense? I haven't checked all the callsites, but it seems like maybe we could have `createDeclResult`, `createKeywordResult`, etc. Could make it easier to understand which (if any) should be used (and where).

rintaro replied:
Ah, that's a good idea 👍

bnbarham commented on lib/IDE/CodeCompletionCache.cpp:
```diff
@@ -209,6 +210,7 @@ static bool readCachedModule(llvm::MemoryBuffer *in,
     auto moduleIndex = read32le(cursor);
     auto briefDocIndex = read32le(cursor);
     auto diagMessageIndex = read32le(cursor);
+    auto filterNameIndex = read32le(cursor);
```
Need to bump `onDiskCompletionCacheVersion`?

bnbarham commented on lib/IDE/CodeCompletionResultPrinter.cpp:
```diff
@@ -450,9 +451,8 @@ void swift::ide::printCodeCompletionResultSourceText(
-    const CodeCompletionResult &Result, llvm::raw_ostream &OS) {
-  auto str = Result.getCompletionString();
+static void printCodeCompletionResultFilterName(
+    const CodeCompletionString *str, llvm::raw_ostream &OS) {
```
Seems like we assume this is not null, `&` instead? There's other functions that take a `CodeCompletionString *` as well, not sure if it applies there too.

rintaro replied:
I agree `&` makes more sense. I just wanted to keep the diff small.

bnbarham replied:
Fair enough, can always do as a follow up. I'm just worried about possible nullptr derefs since this does seem to be null in *some* places.

bnbarham commented on lib/IDE/CodeCompletionResultPrinter.cpp:
```diff
@@ -536,3 +536,12 @@ void swift::ide::printCodeCompletionResultFilterName(
+  SmallString<32> buf;
+  llvm::raw_svector_ostream OS(buf);
+  printCodeCompletionResultFilterName(Str, OS);
+  // Use 'copyCString' to get a null terminated StringRef.
+  return {copyCString(Allocator, buf), buf.size()};
```
Why does it need to be null terminated here?

Also, the usual nitpick of lower vs upper case for the first letter 😅 (quite a few places like this). Honestly at this point though unless we add a formatter and use it in CI, I think I'll just stop pointing it out.

rintaro replied:
I actually prefer 'lowerCamelCase' for local variable 😓

rintaro replied:
I actually prefer `lowerCamelCase` for local variables 😓

bnbarham replied:
Me too. I don't know, we're ostensibly meant to be using the LLVM style guide but... there's no CI for it sooo 🤷

bnbarham commented on include/swift/IDE/CodeCompletionResultPrinter.h:
```diff
@@ -37,8 +40,10 @@ void printCodeCompletionResultTypeNameAnnotated(
-    const CodeCompletionResult &Result, llvm::raw_ostream &OS);
+/// Print 'FilterName' from \p str into memory managed by \p Allocator and
+/// return it.
+llvm::StringRef getCodeCompletionResultFilterName(
```
Nitpick: consistency of allocator first/last (first everywhere else, though personally I'm partial to last :P)

rintaro replied:
I actually prefer 'lowerCamelCase' for local variable 😓

rintaro replied:
Do you mean with `copyString`/`copyCString`? I actually prefer allocator last. I just kept the exiting interface for `copyString`. I can rewrite `copyString`/`copyCString` in this PR if you want.

bnbarham replied:
There was another one as well... I think all the `create` methods.

I also prefer last. If there's only a few places feel free, otherwise we can do it another time.

rintaro replied:
I feel `create()` methods are different. AST's all `create*()` (e.g. `PatternBindingDecl::createImplict`) methods receives `ASTContext &` as the first argument.

bnbarham commented on include/swift/Basic/StringExtras.h:
```diff
@@ -52,12 +52,21 @@ namespace swift {
   /// Determine the part of speech for the given word.
   PartOfSpeech getPartOfSpeech(StringRef word);
 
+  /// Copy \p string to \p Allocator and return it as a \c StringRef .
+  StringRef copyString(llvm::BumpPtrAllocator &Allocator, StringRef string);
+
+  /// Copy \p string to \p Allocator and return it as a null terminated C
+  /// string.
+  const char *copyCString(llvm::BumpPtrAllocator &Allocator, StringRef string);
+
```
Makes sense. I should have done this when I used it in `SwiftSourceDocInfo`. What header in `SwiftSourceDocInfo.cpp` was actually including `CodeCompletion.h` 🤔

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[CodeCompletion] Make all result's string fields null terminated

This is convenient for clients to pass these fields values to C
functions. Introduce NullTerminatedStringRef to guarantee that.

rintaro commented on lib/IDE/CodeCompletionCache.cpp:
```diff
@@ -165,8 +166,8 @@ static bool readCachedModule(llvm::MemoryBuffer *in,
-    auto size = read32le(p);
-    auto str = copyString(*V.Allocator, StringRef(p, size));
+    size_t size = read32le(p);
+    auto str = NullTerminatedStringRef(StringRef(p, size), *V.Allocator);
```
This is used for `CodeCompletionString` too, so their chunk text may needlessly be null-terminated. But there's a high chance of that the chunk text is used as a filter name. So I don't think this increases memory usage much in total.

bnbarham replied:
I'd be more worried about the extra memory from the wrapped StringRef than the single byte from the null termination 🤷

rintaro replied:
I thought memory footprint of `StringRef` object and `NullTerminatedStringRef` object is the same. Just wrapping it increase the memory?

bnbarham replied:
Sorry, you're correct 👍

rintaro commented:
Stress test failure is
```
2 expected stress tester failures not seen
```
(https://bugs.swift.org/browse/SR-14694)
I don't think it's related to this PR.

ahoppen commented on include/swift/Basic/StringExtras.h:
```diff
@@ -465,6 +469,42 @@ bool omitNeedlessWords(StringRef &baseName,
+
+  operator StringRef() const { return Ref; }
+
+  const char *data() const { return Ref.data(); }
```
Suggested doc comment:
```
/// Returns the string as a null-terminated C string.
```

ahoppen commented on include/swift/Basic/StringExtras.h:
```diff
@@ -465,6 +469,42 @@ bool omitNeedlessWords(StringRef &baseName,
+  operator StringRef() const { return Ref; }
+
+  const char *data() const { return Ref.data(); }
+  size_t size() const { return Ref.size(); }
```
Suggested doc comment:
```
The size of the string, excluding the null character.
```

ahoppen commented on include/swift/IDE/CodeCompletionResultPrinter.h:
```diff
@@ -13,12 +13,18 @@
 
 namespace swift {
+
+class NullTerminatedStringRef;
```
Is there a specific reason why you don’t just want to 
```cpp
#include "swift/Basic/StringExtras.h`
```
here? I usually try to avoid forward declarations as much as possible because it makes the code fragile if we should rename/remove `NullTerminatedStringRef` in the future.

bnbarham replied:
I'm actually the opposite 😅. More includes == slower compiles. What do you mean by "fragile" here?

rintaro replied:
FWIW, I confirmed that Xcode's refactoring picked up these (i.e. forward decl and the usages in this header)

ahoppen commented on include/swift/Basic/StringExtras.h:
```diff
@@ -465,6 +469,42 @@ bool omitNeedlessWords(StringRef &baseName,
+
+  constexpr NullTerminatedStringRef(const char *Data) : Ref(Data) {}
+
+  /// Create a null-terminated string copying \p Str into \p A .
```
Nitpick 
```suggestion
  /// Create a null-terminated string, copying \p Str into \p A .
```

ahoppen commented on tools/SourceKit/lib/SwiftLang/CodeCompletionOrganizer.cpp:
```diff
@@ -272,7 +272,7 @@ void CodeCompletionOrganizer::preSortCompletions(
 
      // Sort first by filter name (case-sensitive).
-     if (int primary = (*a)->getName().compare((*b)->getName()))
+     if (int primary = (*a)->getFilterName().compare((*b)->getFilterName()))
```
Unrelated to this PR but in `CodeCompletion.cpp:1484` we sort by case insensitive first and only if that’s equal we compare case-sensitively. Shouldn’t we do the same here? Maybe it would be a good opportunity to add a method to `NullTerminatedStringRef` that compares by case-insensitive followed by case-sensitive.

ahoppen commented on tools/SourceKit/lib/SwiftLang/CodeCompletionOrganizer.cpp:
```diff
@@ -1162,24 +1153,28 @@ Completion *CompletionBuilder::finish() {
         base.getContextFreeResult();
 
+    CodeCompletionString *newStr = contextFreeBase.getCompletionString();
+    NullTerminatedStringRef filterName = contextFreeBase.getFilterName();
```
I think it would avoid confusion if either both of these variables start with with `new` or neither of them. I would suggest either
- `completionString` and `filterName` or
- `newCompletionString` and `newFilterName` (I think I have a slight preference for this one)

ahoppen reviewed (approved):
A few minor suggestions for improvements + comments inline. Otherwise LGTM.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test
